wrapper/pre_copy.py

```diff
@@ -0,0 +1,681 @@
+            nbdkit_cmd.append('user=%s' % self.vmware.user)
+        nbdkit_cmd.extend([
+            'logfile=%s' % STATE.wrapper_log,
+            'logappend=true',
+        ])
+
+        return nbdkit_cmd
+
+    def _start_nbdkits(self, vmware_password_file):
+        paths = []
+        filters = ['cacheextents', 'retry', 'blahaha']
+
+        for filt in filters[:]:
+            try:
+                subprocess.check_call(['nbdkit',
+                                       '--dump-plugin',
+                                       '--filter=' + filt,
+                                       'null'],
+                                      timeout=5)
+                continue
+            except subprocess.TimeoutExpired:
+                filters.remove(filt)
+            except subprocess.CalledProcessError:
+                filters.remove(filt)
+
+        for disk in self.disks:
+            cmd = self._get_nbdkit_cmd(disk, vmware_password_file, filters)
+            logging.debug('Starting nbdkit: %s', cmd)
+            # logging.debug('Logging nbdkit output into: %s', log_path)
+            disk.proc_nbdkit = subprocess.Popen(cmd,
+                                                stdout=subprocess.DEVNULL,
+                                                stderr=subprocess.DEVNULL,
+                                                stdin=subprocess.DEVNULL)
+            paths.append((disk.pidfile, disk.sock))
+
+        logging.debug('Waiting for all nbdkit processes to initialize')
+        endt = time.time() + _TIMEOUT
+        while paths:
+            for path in paths[:]:
+                if os.path.exists(path[0]) and os.path.exists(path[1]):
+                    paths.remove(path)
+            if endt < time.time() or not paths:
+                break
+            time.sleep(.1)
+
+        if paths:
+            raise RuntimeError('Timed out waiting for nbdkits to initialize')
+
+    def _stop_nbdkits(self):
+        for disk in self.disks:
+            if disk.proc_nbdkit is None:
+                continue
+            logging.debug('Stopping nbdkit with pid=%d', disk.proc_nbdkit.pid)
+            disk.proc_nbdkit.terminate()
+            try:
+                disk.proc_nbdkit.wait(timeout=_TIMEOUT)
+            except subprocess.TimeoutExpired:
+                disk.proc_nbdkit.kill()
+                disk.proc_nbdkit.wait()
+            disk.proc_nbdkit = None
+
+    # Returns True/False whether the process is still running
+    def _update_qemu_proc(self, disk, cb_progress):
+        if disk.proc_qemu.poll() is not None:
+            return False
+
+        if cb_progress is not None:
+            try:
+                data = disk.proc_qemu.stdout.read(64)
+                for line in data.rsplit(sep=None, maxsplit=2)[-2:]:
+                    matches = self.qemu_progress_re.search(line)
+                    if matches is not None:
+                        cb_progress(disk, float(matches.group(1)))
+            except IOError as err:
```

Comment:
Actually `IOError` is obsolete and is an alias for `OSError` in Python3. Could you change that, please.